src/Compilers/Core/Rebuild/CompilationOptionsReader.cs

```diff
@@ -136,15 +136,12 @@ public ImmutableArray<MetadataReferenceInfo> GetMetadataReferences()
             ? OutputKind.ConsoleApplication
             : OutputKind.DynamicallyLinkedLibrary;
 
-        public string? GetMainTypeName() => GetMainMethodInfo() is { } tuple
-            ? tuple.MainTypeName
-            : null;
+        // Here we only want to give the caller the main type name if the main method is named "Main" per convention.
+        // If the main method has another name, we have to assume that specifying a main type name won't work.
+        // For example, if the compilation uses top-level statements.
+        public string? GetMainTypeName() => GetMainMethodInfo()?.MainTypeName;
 
-        public string? GetMainMethodName() => GetMainMethodInfo() is { } tuple
-            ? tuple.MainMethodName
-            : null;
-
-        private (string MainTypeName, string MainMethodName)? GetMainMethodInfo()
+        public (string MainTypeName, string MainMethodName)? GetMainMethodInfo()
```

Comment:
This is a `public` method on a `public` type and hence tuples are generally discourage. Don't have to fix in this PR but feel like this will become a `record` before we can "ship" this library.